saleor/dashboard/order/forms.py

```diff
@@ -1,27 +1,267 @@
+                    'Create draft order form error',
+                    'Shipping method is not valid for chosen shipping '
+                    'address')))
+        if errors:
+            raise forms.ValidationError(errors)
+        return self.cleaned_data
+
+    def save(self, commit=True):
+        self.instance.status = OrderStatus.UNFULFILLED
+        if self.instance.user:
+            self.instance.user_email = self.instance.user.email
+        if not self.instance.is_shipping_required():
+            if self.instance.shipping_address:
+                self.instance.shipping_address.delete()
+            self.instance.shipping_method_name = None
+            self.instance.shipping_price = ZERO_TAXED_MONEY
+        return super().save(commit)
+
+
+class OrderCustomerForm(forms.ModelForm):
+    """Set customer details in an order."""
+
+    update_addresses = forms.BooleanField(
+        label=pgettext_lazy(
+            'Update an order with user default addresses',
+            'Set billing and shipping address in order to customer defaults'),
+        initial=True, required=False)
+    user = AjaxSelect2ChoiceField(
+        queryset=User.objects.all(),
+        fetch_data_url=reverse_lazy('dashboard:ajax-users-list'),
+        required=False)
+
+    class Meta:
+        model = Order
+        fields = ['user', 'user_email']
+        labels = {
+            'user': pgettext_lazy('Order customer', 'User'),
+            'user_email': pgettext_lazy(
+                'Order customer email',
+                'Email')}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.instance.user:
+            self.fields['user'].set_initial(
+                self.instance.user, label=self.instance.user.ajax_label)
+
+    def clean(self):
+        cleaned_data = super().clean()
+        user_email = cleaned_data.get('user_email')
+        user = cleaned_data.get('user')
+        if user and user_email:
+            raise forms.ValidationError(pgettext_lazy(
+                'Edit customer details in order form error',
+                'An order can be related either with an email or an existing '
+                'user account'))
+        return self.cleaned_data
+
+    def save(self):
+        super().save()
+        if self.cleaned_data.get('update_addresses'):
+            update_order_with_user_addresses(self.instance)
+        return self.instance
+
+
+class OrderRemoveCustomerForm(forms.ModelForm):
+    """Remove customer data from an order."""
+
+    class Meta:
+        model = Order
+        fields = []
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def save(self):
+        customer = self.instance.user
+        self.instance.user = None
+        self.instance.user_email = ''
+        super().save()
+
+        if customer:
+            if (
+                customer.default_billing_address ==
```

Comment:
Why are we doing this? Maybe some comment would be helpful, for one going back to this after some time

Reply:
Cause if we set billing and shipping address in an order to default user addresses before, we want to unset this addresses when removing customer. But if custom changes were made, we leave it, as it might have been intended. I'll add a comment or maybe move whole logic of removing customer to separate function, marked with docstring.